tunnistamo/tests/test_suomifi_authentication.py

```diff
@@ -0,0 +1,353 @@
+from freezegun import freeze_time
+from oidc_provider.models import Client, Token
+from onelogin.saml2.utils import OneLogin_Saml2_Utils as SAMLUtils
+from social_django.models import UserSocialAuth
+
+from users.models import Application, LoginMethod, OidcClientOptions
+from users.views import LoginView
+
+SERVER_NAME = 'tunnistamo.test'
+CLIENT_NAME = 'Test Client'
+CLIENT_ID = 'test_client'
+REDIRECT_URI = 'https://tunnistamo.test/redirect_uri'
+ID_TOKEN = {'aud': 'test_client'}
+ID_TOKEN_JWT = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJ0ZXN0X2NsaWVudCJ9'
+ID_TOKEN_JWT_INVALID = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJJTlZBTElEIn0'
+RELAY_STATE = 'SUOMIFI_RELAY_STATE'
+
+
+@pytest.fixture
+def django_client(request):
+    from django.test.client import Client
+    return Client(SERVER_NAME=SERVER_NAME)
+
+
+@pytest.fixture
+def fixed_saml_id(monkeypatch):
+    monkeypatch.setattr(SAMLUtils, 'generate_unique_id', lambda: 'MESSAGE_ID_FOR_TEST')
+
+
+def create_test_oidc_client():
+    test_client = Client.objects.create(
+        name=CLIENT_NAME,
+        client_id=CLIENT_ID
+    )
+    test_client.redirect_uris = (REDIRECT_URI,)
+    test_client.post_logout_redirect_uris = (REDIRECT_URI,)
+    test_client.save()
+
+    login_method = LoginMethod.objects.create(
+        provider_id='suomifi',
+        name='suomi.fi.test'
+    )
+
+    client_options = OidcClientOptions.objects.create(
+        oidc_client=test_client
+    )
+    client_options.login_methods.add(login_method)
+    client_options.save()
+
+    return test_client
+
+
+def create_noncompliant_test_provider():
+    test_provider = Application.objects.create(
+        client_id=CLIENT_ID,
+        client_type=Application.CLIENT_CONFIDENTIAL,
+        authorization_grant_type=Application.GRANT_AUTHORIZATION_CODE,
+        name=CLIENT_NAME
+    )
+    test_provider.redirect_uris = REDIRECT_URI
+    login_method = LoginMethod.objects.create(
+        provider_id='suomifi',
+        name='suomi.fi.test'
+    )
+    test_provider.login_methods.add(login_method)
+    test_provider.save()
+
+
+def load_file(filename):
+    with open(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                           'data',
+                           filename), 'rb') as xmlfile:
+        return xmlfile.read()
+
+
+@pytest.mark.django_db
+@freeze_time('2019-01-01 12:00:00', tz_offset=2)
+def test_suomifi_metadata(django_client):
+    create_test_oidc_client()
+    metadata_url = reverse('auth_backends:suomifi_metadata')
+    metadata_response = django_client.get(metadata_url)
+    expected_metadata = load_file('suomifi_metadata.xml')
+    assert metadata_response.status_code == 200
+    assert metadata_response.get('content-type') == 'text/xml'
+    assert metadata_response.content == expected_metadata
+
+
+@pytest.mark.django_db
+@freeze_time('2019-01-01 12:00:00', tz_offset=2)
+def test_suomifi_login_request(django_client, fixed_saml_id):
+    '''Suomi.fi use case #1'''
```

Comment:
Where are these use cases documented and numbered? I'll store the reference in an appropriate place.

Reply:
https://palveluhallinta.suomi.fi/fi/sivut/tunnistus/kayttoonotto/asiakastestiymparisto